### DISCUSSION TITLE
Questions about converting networks from lava-dl to lava

### DISCUSSION THREAD

**Discussion by fangwei123456** [lava-nc/lava#245] [Q&A] [answered]
`lava.lib.dl.slayer.block` provides a simple method to define deep SNN and conver SNN to process in Lava by `lava.lib.dl.next`. I build a toy network and test it:

```python
from lava.magma.core.run_configs import Loihi1SimCfg
from lava.magma.core.run_conditions import RunSteps
from lava.proc.io.sink import RingBuffer as ReceiveProcess
from lava.proc.io.source import RingBuffer as SendProcess
import h5py
import torch
from lava.lib.dl import netx, slayer
filename ='./lava_net.net'
tau = 2.
scale = 1 << 6
in_features = 8
out_features = 1
N = 1
T = 8
neuron_param = {
            'threshold': 1.,
            'current_decay': 1.,
            'voltage_decay': 1. / tau,
            'tau_grad': 1, 'scale_grad': 1, 'scale': scale,
            'norm': None, 'dropout': None,
            'shared_param': True, 'persistent_state': False, 'requires_grad': False,
            'graded_spike': False
        }
with torch.no_grad():
    net = slayer.block.cuba.Dense(neuron_param, in_features, out_features)
    torch.nn.init.constant_(net.synapse.weight.data, 0.25)
    x = torch.rand([N, in_features, T])
    x = (x > 0.5).float()
    y = net(x)
    print('lava-dl output:\n', y.numpy()[0])
    h5_h = h5py.File(filename, 'w')
    h5_layer = h5_h.create_group('layer')
    net.export_hdf5(h5_layer.create_group(f'{0}'))

    net_x = netx.hdf5.Network(net_config=filename)
    source = SendProcess(x[0])
    sink = ReceiveProcess(shape=net_x.out.shape, buffer=T * 2)
    source.s_out.connect(net_x.inp)
    net_x.out.connect(sink.a_in)
    run_condition = RunSteps(num_steps=T)
    run_config = Loihi1SimCfg(select_tag='fixed_pt')
    net_x.run(condition=run_condition, run_cfg=run_config)
    output = sink.data.get()
    net_x.stop()
    print('lava output:\n', output)
```

The outputs are:
```
lava-dl output:
 [[0. 1. 0. 1. 1. 1. 1. 1.]]
lava output:
 [[0. 0. 1. 1. 0. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0.]]
```

I find that the two networks have different outputs (`0. 1. 0. 1. 1. 1. 1. 1.` is not in `0. 0. 1. 1. 0. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0.`). Is the difference expected?

**Comment by bamsumit** [comment, marked as answer]
Hi @fangwei123456, thanks for reaching out to us. The neurons in slayer and lava should match except for some esoteric cases. What you have encountered is a specific discrepancy regarding the spike condition. A fix to this issue has been pushed in [lava-dl PR #78](https://github.com/lava-nc/lava-dl/pull/78). Once it is merged, the difference should be resolved and the outputs should match 1:1

The PR also includes a test inspired by your code snippet. Thanks for uncovering such issues. We highly appreciate it.

**Comment by fangwei123456** [comment]
Hi @bamsumit , I have another question about it. If a comment `x = (x > 0.5).float()`, then the input will be not binary, and I will get different outputs from lava/lava-dl. Does snn in lava only supports for binary input? And if I set `graded_spike=True` , I find in the source codes:

```
spikes = ((voltage >= threshold) * voltage / scale).to(dtype)
```

Does it means the spikes are also 8-bits, which is same with `voltage` ?

**Reply by bamsumit** [reply]
If `graded_spike = True`, we can send spikes with magnitude. This is a feature supported by Loihi 2 and hence lava-dl also supports it.

In the code above, if the voltage exceeds the threshold, a spike with a value equal to the voltage is sent.
